x/ibc/23-commitment/types/types.proto

```diff
@@ -38,7 +38,7 @@ message MerklePath {
 message MerkleProof {
   option (gogoproto.equal) = true;
 
-  tendermint.crypto.merkle.Proof proof = 1;
+  repeated ics23.CommitmentProof proofs = 1;
```

Comment:
Yes that should work, the key is to not use merkle.ProofRuntime to verify this proof since that will use any of the decoders in any order to decode and verify the proof.

Instead we need to take the list of specs implemented either as ([]ics23.ProofSpec, or []string) and then verify that each proof in the chain is as we expect

Reply:
As the last comment mention, the simplest first step here for correctness is to:

(1) add the []string proofTypes argument
(2) check the `tendermint.crypto.merkle.Proof` proofs are present in the right number and types
(3) then call into merkle.ProofRuntime

This still leaves the proof runtime as is, but shores it up to work.

Reply:
Once this is correct, you can start a larger refactoring about removing merkle.ProofOp. But I would do so everywhere in the codebase, and only after accepted by the core sdk team. With just the change I mentioned, this is a bit ugly (semi-needless abstraction), but has the same correctness guarantees you are aiming for. And you can embed SimpleProof as well as an ics23 proof and have a "verifiable by non-go-code" ibc packets